File: Exercicios/Python/Lista1/Ex2.py
# SSC0103 - Programação Orientada a Objetos
# Lista de exercícios #1 (Python)
# Exercício 02
# 
# Ana Lívia Ruegger Saldanha (N.USP 8586691)

'''
Este programa lê do teclado os coeficientes de uma equação de segundo grau
e apresenta sua solução.
'''


import math

def handles_exception():
    ''' Esta função trata exceções na leitura de um valor inteiro do teclado.
    Caso a leitura não seja válida, pede que o usuário digite novamente até
    obter um valor válido.
    '''
    read_int = False
    n = -1 # Iniciando a variável com um valor qualquer.

    while not read_int:
        try:
            n = int(input())
            read_int = True
        except:
            print("Você deve digitar um valor inteiro!\nDigite novamente.")
    
    return n

print("Digite os coeficientes da equação de segundo grau:")
a = handles_exception()
b = handles_exception()
c = handles_exception()

delta = b ** 2 - 4 * a * c

if delta < 0:
    print("Esta equação não possui raízes reais.")
else:
    x1 = (-b - math.sqrt(delta)) / (2 * a)
    if delta == 0:
        print("A raiz dupla desta equação é", x1)
    else:
        x2 = (-b + math.sqrt(delta)) / (2 * a)
        print("As raízes da equação são", x1, "e", x2)

File: Exercicios/Java/Lista1/src/Ex2.java
/**
 * SSC0103 - Programação Orientada a Objetos
 * Lista de exercícios #1 (Java)
 * Exercício 02
 * 
 * @author Ana Lívia Ruegger Saldanha (N.USP 8586691)
 *
 */
public class Ex2 {
	
	/**
	 * Este método lê do teclado os coeficientes de uma equação de
	 * segundo grau e apresenta sua solução.
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		System.out.println("Digite os coeficientes da equação:");
		int a, b, c;
		
		a = handlesExceptionInt();
		b = handlesExceptionInt();
		c = handlesExceptionInt();
		
		int delta = (int)(Math.pow(b, 2)) - 4 * a * c;
		
		if (delta < 0) {
			System.out.println("Esta equação não possui raízes reais.");
		} else {
			double x1 = (-b - Math.sqrt(delta)) / (2 * a);
			if (delta == 0) {
				System.out.printf("A raiz dupla desta equação é %f\n", x1);
			} else {
				double x2 = (-b + Math.sqrt(delta)) / (2 * a);
				System.out.printf("As raízes da equação são %f e %f\n", x1, x2);
			}
		}
	}
	
	/**
	 * Este método trata exceções na leitura de um inteiro do
	 * teclado. Caso a leitura não seja válida, pede que o usuário
	 * digite novamente até obter um valor válido.
	 * 
	 * @return inteiro lido.
	 */
	public static int handlesExceptionInt() {
		boolean read_int = false;
		int n = -1; // Iniciando a variável com um valor qualquer.
		
		while (!read_int) {
			try {
				n = EntradaTeclado.leInt();
				read_int = true;
			} catch (Exception e) {
				System.out.println("Você deve digitar um inteiro!");
			}
		}
		
		return n;
	}

}
